src/shm/pool/multi.rs

```diff
@@ -0,0 +1,398 @@
+    #[error("buffer is currently used")]
+    InUse,
+    #[error("buffer is overlapping another")]
+    Overlap,
+    #[error("buffer could not be found")]
+    NotFound,
+}
+
+/// This pool manages buffers associated with keys.
+/// Only one buffer can be attributed to a given key.
+#[derive(Debug)]
+pub struct MultiPool<K> {
+    buffer_list: Vec<BufferSlot<K>>,
+    pub(crate) inner: RawPool,
+}
+
+#[derive(Debug, thiserror::Error)]
+struct BufferSlot<K> {
+    free: Arc<AtomicBool>,
+    size: usize,
+    used: usize,
+    offset: usize,
+    buffer: Option<wl_buffer::WlBuffer>,
+    key: K,
+}
+
+impl<K> MultiPool<K> {
+    pub(crate) fn new(inner: RawPool) -> Self {
+        Self { inner, buffer_list: Vec::new() }
+    }
+
+    /// Resizes the memory pool, notifying the server the pool has changed in size.
+    ///
+    /// The wl_shm protocol only allows the pool to be made bigger. If the new size is smaller than the
+    /// current size of the pool, this function will do nothing.
+    pub fn resize(&mut self, size: usize) -> io::Result<()> {
+        self.inner.resize(size)
+    }
+
+    /// Removes the buffer with the given key from the pool and rearranges the others.
+    pub fn remove<Q>(&mut self, key: &Q)
+    where
+        Q: PartialEq,
+        K: std::borrow::Borrow<Q>,
+    {
+        if let Some((i, buf)) =
+            self.buffer_list.iter().enumerate().find(|(_, slot)| slot.key.borrow().eq(key))
+        {
+            let mut offset = buf.offset;
+            if let Some(buf) = self.buffer_list.remove(i).buffer {
+                buf.destroy();
+            }
+            for buf_slot in &mut self.buffer_list {
+                if buf_slot.offset > offset && buf_slot.free.load(Ordering::Relaxed) {
+                    if let Some(buffer) = buf_slot.buffer.take() {
+                        buffer.destroy();
+                    }
+                    std::mem::swap(&mut buf_slot.offset, &mut offset);
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Insert a buffer into the pool.
+    ///
+    /// The parameters are:
+    ///
+    /// - `width`: the width of this buffer (in pixels)
+    /// - `height`: the height of this buffer (in pixels)
+    /// - `stride`: distance (in bytes) between the beginning of a row and the next one
+    /// - `key`: a borrowed form of the stored key type
+    /// - `format`: the encoding format of the pixels.
+    pub fn insert<Q>(
+        &mut self,
+        width: i32,
+        stride: i32,
+        height: i32,
+        key: &Q,
+        format: wl_shm::Format,
+    ) -> Result<usize, PoolError>
+    where
+        K: Borrow<Q>,
+        Q: PartialEq + ToOwned<Owned = K>,
+    {
+        let mut offset = 0;
+        let mut found_key = false;
+        let size = (stride * height) as usize;
+        let mut index = Err(PoolError::NotFound);
+        let bpp = (stride as f32 / width as f32).ceil() as usize;
```

Comment:
With a 5x5 ARGB8888 buffer whose rows are aligned to 8 bytes, you have `width = 5, stride = 24, bpp = 5` - likely not what is intended.

Reply:
I just tried creating a 5x5 ARGB8888 buffer and the stride is 20. It's not aligned to 8 bytes from what I observed.

Reply:
Correct, if your graphics library only requires 4-byte alignment then it will have a stride of 20.  However, if your graphics library uses 8- or 16-byte alignment for its buffers, it could be 24 or even 32.  This could be done in order to take advantage of 128-bit SSE instructions in rendering without needing to mask off bits at the end.